File: 1501~2000/1631. Path With Minimum Effort/main.py
"""
Source: https://rb.gy/xibtg
Date: 2023/8/21
Skill:
Ref:
Runtime: 519 ms, faster than 97.98%
Memory Usage: 17.64 MB, less than 88.70%
Time complexity:
Space complexity:
Constraints:
    1 <= rows, columns <= 100
    1 <= heights[i][j] <= 10 ** 6
"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right
from functools import cache


class Solution:
    def minimumEffortPath(self, heights: List[List[int]]) -> int:
        m, n, res, heap = len(heights), len(heights[0]), 0, [[0, 0, 0]]
        dirs = [0, 1, 0, -1, 0]
        heapify(heap)
        while heap:
            diff, x, y = heappop(heap)
            if heights[x][y] == -1:
                continue
            res = max(res, diff)
            if x == m - 1 and y == n - 1:
                return res
            for i in range(4):
                next_x, next_y = x + dirs[i], y + dirs[i + 1]
                if 0 <= next_x < m and 0 <= next_y < n and heights[next_x][next_y] != -1:
                    heappush(heap, [abs(heights[x][y] - heights[next_x][next_y]), next_x, next_y])
            heights[x][y] = -1

        return res


if __name__ == "__main__":
    s = Solution()
    heights = [[10, 8], [10, 8], [1, 2], [10, 3], [1, 3], [6, 3], [5, 2]]
    res = s.minimumEffortPath(heights)

File: 1501~2000/1631. Path With Minimum Effort/main.cpp
/**
 * Source: https://rb.gy/xibtg
 * Date: 2023/9/16
 * Skill:
 * Runtime: 110 ms, faster than 69.39% of C++ online submissions
 * Memory Usage: 19.13 MB, less than 96.78% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */

#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>

using namespace std;

#define ll long long
typedef pair<int, int> pairs;
typedef unsigned long long ULL;
typedef pair<ULL, ULL> PULL;


class Solution {
public:
    int minimumEffortPath(vector<vector<int>>& heights) {
        int m = heights.size(), n = heights[0].size(), dir[5] = {0, 1, 0, -1, 0};
        vector<bool> visited(m * n, false);
        priority_queue<pairs, vector<pairs>, greater<pairs>> pq;
        pq.push({0, 0});
        while (!pq.empty()) {
            auto [effort, pos] = pq.top(); pq.pop();
            if (pos == m * n - 1)
                return effort;
            if (visited[pos])
                continue;
            visited[pos] = true;
            int x = pos / n, y = pos % n;
            for (int i = 0; i < 4; i++) {
                int xx = x + dir[i], yy = y + dir[i + 1], pos2 = xx * n + yy;
                if (xx < 0 || xx >= m || yy < 0 || yy >= n || visited[pos2])
                    continue;
                pq.push({max(effort, abs(heights[x][y] - heights[xx][yy])), pos2});
            }
        }
        return 0;
    }
};


int main() {
    Solution s;
    vector<int> nums{1,1,3,3};
    vector<vector<int>> arr{{0,1,1}, {1,1,0}, {1,1,0}};
}